fms/testing/fms_hf_equivalence.py

```diff
@@ -0,0 +1,314 @@
+            },
+            "sequence_info": {
+                "sequences_with_extra_tokens": self.sequences_with_extra_tokens,
+                "sequences_with_less_tokens": self.sequences_with_less_tokens,
+                "sequences_with_different_tokens": self.sequences_with_different_tokens,
+            },
+        }
+
+
+def compute_diff(generated_tokens_dir):
+    hf_file = tokens_file_path(generated_tokens_dir, "hf")
+    fms_file = tokens_file_path(generated_tokens_dir, "fms")
+    _compute_diff(hf_file, fms_file)
+
+
+def _compute_diff(file1, file2):
+    """
+    Computes the difference in the tokens stored in two files, produced by the function 'generate_and_write_tokens' and prints the summary.
+    """
+    dict1 = torch.load(file1, map_location=torch.device(_decice_for_diff))
+    dict2 = torch.load(file2, map_location=torch.device(_decice_for_diff))
+
+    if len(dict1) != len(dict2):
+        logging.warn(f"The number of entries differ")
+        return
+
+    summary = DiffSummary()
+
+    for k1, v1 in dict1.items():
+        v2 = dict2[k1]
+
+        if k1.startswith("input"):
+            assert torch.equal(v1, v2)
+            continue
+
+        prompt_number = k1.split("-", 1)[-1].strip()
+        summary.total_sequences += 1
+        v1_size = v1.size()[0]
+        v2_size = v2.size()[0]
+        size_diff = v1_size - v2_size
+
+        if size_diff > 0:
+            summary.less_tokens += size_diff
+            summary.sequences_with_less_tokens.append(prompt_number)
+            v1 = v1[:v2_size]
+
+        elif size_diff < 0:
+            summary.extra_tokens += -size_diff
+            summary.sequences_with_extra_tokens.append(prompt_number)
+            v2 = v2[:v1_size]
+
+        summary.total_tokens += v1_size
+
+        diff = torch.sub(v1, v2)
+        num_diff = torch.count_nonzero(diff).item()
+        summary.different_tokens += num_diff
+        if num_diff > 0:
+            summary.sequences_with_different_tokens.append(prompt_number)
+
+    logging.info(f"diff summary = {json.dumps(summary.toJson(), indent=2)}")
+
+
+def main(argv):
+    logging.basicConfig(
+        level=logging.DEBUG,
+        format="%(asctime)s - %(levelname)s - %(message)s",
+        datefmt="%Y-%m-%d %H:%M:%S",
+    )
+
+    parser = argparse.ArgumentParser(
+        description="Utility functions for comparing outputs of two implementations of a model"
+    )
+    subparsers = parser.add_subparsers(dest="subcommand")
+
+    fms = subparsers.add_parser("run_fms_model", help="Run fms model")
+    fms.add_argument(
+        "-d", "--dtype", type=str, required=True, choices=["float16", "float32"]
+    )
+    fms.add_argument("-a", "--model_arch", type=str, required=True)
+    fms.add_argument("-v", "--model_variant", type=str, required=True)
+    fms.add_argument("-p", "--model_path", type=str, required=True)
```

Comment:
at some point we should refactor to make the model creation argument construction shared across our scripts, but for now might want to copy/paste to at least make the usage consistent. E.g. in other scripts we have 'architecture', 'variant', 'model_path', 'model_source'

Reply:
Done